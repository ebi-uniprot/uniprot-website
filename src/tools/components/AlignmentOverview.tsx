import React, { FC, useCallback, memo } from 'react';

import useCustomElement from '../../shared/hooks/useCustomElement';
import { SegmentTrackData } from '../utils/sequences';

type AlignmentOverviewProps = {
  height: string;
  data: SegmentTrackData[][];
  length: number;
  highlight: string;
};

type AlignmentOverviewTrackProps = {
  height: number;
  data: SegmentTrackData[];
  length: number;
  highlight: string;
};

const AlignmentOverviewTrack: FC<AlignmentOverviewTrackProps> = ({
  data,
  highlight,
  length,
  height,
}) => {
  const ceDefined = useCustomElement(
    /* istanbul ignore next */
    () => import(/* webpackChunkName: "protvista-track" */ 'protvista-track'),
    'protvista-track'
  );

  const setTrackData = useCallback(
    (node): void => {
      if (node && ceDefined) {
        // eslint-disable-next-line no-param-reassign
        node.data = data;
      }
    },
    [data, ceDefined]
  );

  return (
    <protvista-track
      height={height}
      ref={setTrackData}
      length={length}
      layout="non-overlapping"
      highlight={highlight}
    />
  );
};

const AlignmentOverview: FC<AlignmentOverviewProps> = memo(
  ({ height, data, length, highlight }) => {
    if (!data || data.length < 1) {
      return null;
    }

    const singleTrackHeight = Math.floor(parseInt(height, 10) / data.length);

    return (
      <div>
        {data.map((trackData, index) => {
          return (
            <AlignmentOverviewTrack
              data={trackData}
              height={singleTrackHeight}
              length={length}
              highlight={highlight}
              // eslint-disable-next-line react/no-array-index-key
              key={index}
            />
          );
        })}
      </div>
    );
  }
);

export default AlignmentOverview;
