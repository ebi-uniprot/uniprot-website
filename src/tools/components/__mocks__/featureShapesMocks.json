{
    "BLAST": {
        "length": 751,
        "highlight": "41:125",
        "data": [
            [{
                "start": 40,
                "end": 40,
                "shape": "line",
                "color": "#014371",
                "locations": [{
                    "fragments": [{
                        "start": 40,
                        "end": 40
                    }]
                }]
            }, {
                "start": 41,
                "end": 267,
                "color": "#014371",
                "locations": [{
                    "fragments": [{
                        "start": 41,
                        "end": 267
                    }]
                }]
            }, {
                "start": 268,
                "end": 582,
                "shape": "line",
                "color": "#014371",
                "locations": [{
                    "fragments": [{
                        "start": 268,
                        "end": 582
                    }]
                }]
            }],
            [{
                "start": 0,
                "end": 40,
                "shape": "line",
                "color": "#014371"
            }, {
                "start": 41,
                "end": 267,
                "color": "#014371"
            }, {
                "start": 268,
                "end": 751,
                "shape": "line",
                "color": "#014371"
            }]
        ]
    }
}
