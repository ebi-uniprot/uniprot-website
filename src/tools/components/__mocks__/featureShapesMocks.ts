import { Feature } from '@nightingale-elements/nightingale-track';

type T = {
  length: number;
  highlight: string;
  data: Feature[][];
};
const mock: T = {
  length: 751,
  highlight: '41:125',
  data: [
    [
      {
        accession: '1',
        start: 40,
        end: 40,
        shape: 'line',
        color: '#014371',
        locations: [
          {
            fragments: [
              {
                start: 40,
                end: 40,
              },
            ],
          },
        ],
      },
      {
        accession: '2',
        start: 41,
        end: 267,
        color: '#014371',
        locations: [
          {
            fragments: [
              {
                start: 41,
                end: 267,
              },
            ],
          },
        ],
      },
      {
        accession: '3',
        start: 268,
        end: 582,
        shape: 'line',
        color: '#014371',
        locations: [
          {
            fragments: [
              {
                start: 268,
                end: 582,
              },
            ],
          },
        ],
      },
    ],
    [
      {
        accession: '4',
        start: 0,
        end: 40,
        shape: 'line',
        color: '#014371',
      },
      {
        accession: '5',
        start: 41,
        end: 267,
        color: '#014371',
      },
      {
        accession: '6',
        start: 268,
        end: 751,
        shape: 'line',
        color: '#014371',
      },
    ],
  ],
};

export default mock;
