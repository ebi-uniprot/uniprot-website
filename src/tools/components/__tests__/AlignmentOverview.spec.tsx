import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import renderWithRedux from '../../../shared/__test-helpers__/RenderWithRedux';

import AlignmentOverview from '../AlignmentOverview';

import { BLAST as mockData } from '../__mocks__/featureShapesMocks.json';

describe('AlignmentOverview', () => {
  describe('BLAST', () => {
    let rendered;
    beforeEach(() => {
      rendered = renderWithRedux(
        <AlignmentOverview
          data={mockData.data}
          height="10"
          length={mockData.length}
          highlight={mockData.highlight}
        />
      );
    });

    it('should render', () => {
      const { asFragment } = rendered;
      expect(asFragment()).toMatchSnapshot();
    });

    it('should not render when data is missing from the props', () => {
      const { asFragment } = renderWithRedux(
        <AlignmentOverview
          data={[]}
          height="10"
          length={mockData.length}
          highlight={mockData.highlight}
        />
      );
      expect(asFragment().textContent).toEqual('');
    });
  });
});
