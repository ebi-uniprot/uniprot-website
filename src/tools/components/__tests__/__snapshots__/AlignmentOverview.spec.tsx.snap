// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlignmentOverview BLAST should render 1`] = `
<DocumentFragment>
  <div>
    <protvista-track
      height="5"
      highlight="41:125"
      layout="non-overlapping"
      length="751"
    />
    <protvista-track
      height="5"
      highlight="41:125"
      layout="non-overlapping"
      length="751"
    />
  </div>
</DocumentFragment>
`;
